Shell completions for Cargo

yodaldevoid opened the pull request:
This is a cleanup of a prior PR (#1425) by @ricvelozo.

- Adds the ability to output shell completions for cargo using the command `rustup completions <shell> cargo`
- Updates the completion instructions to mention and suggest placing bash completions in the home directory.

ricvelozo committed:
Suggest home directory for bash completion

ricvelozo committed:
Include shell completion for Cargo

ricvelozo committed:
Update the README

yodaldevoid committed:
Move completion command handler to separate function

Signed-off-by: Gabriel Smith <ga29smith@gmail.com>

yodaldevoid committed:
Add optional "command" argument to rustup completion

This "command" argument allows rustup to output the completion information of
more than just rustup. In this case it adds a cargo completion script.

Signed-off-by: Gabriel Smith <ga29smith@gmail.com>

yodaldevoid committed:
For cargo completion script, output line to source real script

This allows the cargo script for the current default toolchain to be called,
and if the script is ever updated the user won't have to maually update their
completion script after updating Rust.

Signed-off-by: Gabriel Smith <ga29smith@gmail.com>

yodaldevoid committed:
Add information to the help output about bash completions

Signed-off-by: Gabriel Smith <ga29smith@gmail.com>

kinnison commented:
Hi @yodaldevoid 

Thanks for your contribution.  Could you possibly tidy up the commit sequence (e.g. fold the fixes to formatting into the requisite commits instead of having a 'run rustfmt' commit) to make it easier to review?

I've put this PR on my list to watch but I'd like to see things a little tidier before I review.

I look forward to your changes.

D

yodaldevoid commented:
Hello @kinnison 

Thanks for getting to looking over this. I've tidied up the rustfmt commit, but no other commits jumped out at me as being appropriate to tidy up. Is there anything else you would like cleaned up?

I should note that I wanted to keep @ricvelozo's commits mostly intact. They could get merged together with other commits, to condense everything, but I wanted to preserve his contribution.

kinnison commented:
Fair enough, I can appreciate wanting to maintain that.  I'll put this on my list to review as soon as I can (likely within a day or two)

bors commented:
:umbrella: The latest upstream changes (presumably #1630) made this pull request unmergeable. Please resolve the merge conflicts.

yodaldevoid commented:
@kinnison Just wanted to ping you to make sure this hasn't fallen by the wayside.

kinnison commented on README.md:
````diff
@@ -67,7 +67,7 @@ but the gist is as simple as using one of the following:
 ```
 # Bash
-$ rustup completions bash > /etc/bash_completion.d/rustup.bash-completion
+$ rustup completions bash > ~/.local/share/bash_completion/completions/rustup
````
Does bash pick this up by default?  if so, from what version?

yodaldevoid replied:
`bash-completions` (the package that adds completions to `bash` and is a de-facto standard to the point that people don't realize that `bash` does not have this functionality built in) added this with this commit scop/bash-completion@c89dcbbd5510876f6304ef10806b00cc9fda19dc, which was included in release 2.2 on March 3rd, 2016 (https://github.com/scop/bash-completion/releases/tag/2.2).

kinnison commented on src/rustup-cli/rustup_mode.rs:
```diff
@@ -1038,3 +1098,52 @@ fn set_default_host_triple(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
+        match s {
+            _ if s.eq_ignore_ascii_case("rustup") => Ok(CompletionCommand::Rustup),
+            _ if s.eq_ignore_ascii_case("cargo") => Ok(CompletionCommand::Cargo),
+            _ => Err(String::from("[valid values: rustup, cargo]")),
```
I dislike that `variants()` enumerates the possible strings, *this* `match` enumerates the strings, **and** this error string enumerates the strings, which match the completion commands.

I'd almost prefer:

```rust
static COMPLETIONS: &[(&'static str, CompletionCommand)] = &[
    ("rustup", CompletionCommand::Rustup),
    ("cargo", CompletionCommand::Cargo),
];
```

combined with a `from_str` which uses that array, and a `variants()` which maps it into a list of strings, and then this error can be constructed from that list too.  All in all, it just feels a little too denormalised for my liking.

kinnison commented on src/rustup-cli/rustup_mode.rs:
```diff
@@ -1038,3 +1098,52 @@ fn set_default_host_triple(cfg: &Cfg, m: &ArgMatches<'_>) -> Result<()> {
+        CompletionCommand::Cargo => {
+            let script = match shell {
+                Shell::Bash => Some("/etc/bash_completion.d/cargo"),
+                Shell::Zsh => Some("/share/zsh/site-functions/_cargo"),
```
What about other shells, does `cargo` only support bash and zsh?  If so, should we report an error here too if a different shell is selected?

yodaldevoid replied:
`cargo` currently only has completion scripts for `bash` and `zsh`. I agree that we should report an error if a user specifies a shell that is not yet supported. I will add a message printed out on stderr.

kinnison reviewed (changes requested):
First up, thank you for submitting a PR and contributing to `rustup`.

This is not a bad PR, though there are a number of points which need cleanup. The big thing is that there's a distinct lack of tests.

Could we please have some basic tests for the new command you're adding?

Thanks,

D.

ehuss commented:
FWIW, I'm proposing a new approach for completions for cargo at https://github.com/rust-lang/cargo/issues/6645.  It may not have any impact here, but just wanted to make you aware.

yodaldevoid commented:
@ehuss I'll keep an eye on that discussion, and when/if it gets implemented if no one else does it I'll make sure to submit a PR moving to the new system for newer versions of `cargo`.

dwijnand commented:
Thanks, @ehuss, I've been meaning to cross link that here.

bors commented:
:umbrella: The latest upstream changes (presumably #1704) made this pull request unmergeable. Please resolve the merge conflicts.

yodaldevoid committed:
Print an error message for unsupported cargo completions

yodaldevoid committed:
Use term2 for completion script outputs

yodaldevoid committed:
Hold all programs with completions in an array

This removes the duplication between the `variants` and `from_str`
implementations.

yodaldevoid committed:
cli: Rework completion error handling

Moves the error reporting from the completion function to the standard
error handling mechanism.

CompletionCommand had to be made fully public, not that this matters for
a binary target. That said, this could have been prevented by
transforming the CompletionCommand into its String form at the error
site rather than at the reporting site. It was chosen to go with this
method as stringly-typed interfaces are generally bad.

yodaldevoid committed:
test: mock: Add expect_ok_eq function

Compares two different command invocations and panics if their return
code and stdout/stderr don't match.

yodaldevoid committed:
test: Added tests for shell completion script support

yodaldevoid commented:
@kinnison Sorry about the long time since the last update. As the famous quote goes, "Life... finds a way to always get in the way of projects."

I believe I have addressed all requested changes. Please let me know if there is anything else you would like modified.

kinnison commented:
Thank you for making the updates, I'll endeavour to review this, likely tomorrow.

yodaldevoid commented:
@kinnison I hate to be a hypocrite in this regard, but I wanted to ping you about this PR as it has been some time.

kinnison commented:
> @kinnison I hate to be a hypocrite in this regard, but I wanted to ping you about this PR as it has been some time.

No, that's very fair, I needed a kick up the backside it seems the past two weeks have been busy and somehow this slipped off my TODO without me noticing :(

kinnison commented:
This looks reasonable to me, I'm grateful for the tests too.  I wonder if the user ought to be told about how the completions script they're given may vary with the default toolchain; but ultimately I think that's not worth blocking this for.

pickfire commented:
Are there fish completions yet?

tesuji commented:
If you want to contribute, please help with https://github.com/rust-lang/cargo/issues/6645 .
